padertorch/ops/losses/source_separation.py

```diff
@@ -105,18 +105,194 @@ def pit_loss(
+        tensor(1.)
+
+        >>> T, K, F = 4, 2, 5
+        >>> estimate, target = torch.ones(T, K, F), torch.zeros(T, F, dtype=torch.int64)
+        >>> pit_loss_from_pair_wise(pair_wise_loss(estimate, target, 1, loss_fn=torch.nn.functional.cross_entropy), reduction='sum')
+        tensor(0.6931)
+        >>> pit_loss(estimate, target, 1, loss_fn=torch.nn.functional.cross_entropy)
+        tensor(0.6931)
+
+        >>> T, K, F = 4, 2, 5
+        >>> estimate, target = torch.ones(K, F, T), torch.zeros(K, F, T)
+        >>> pit_loss_from_pair_wise(pair_wise_loss(estimate, target, 0))
+        tensor(1.)
+
+        >>> T, K, F = 4, 2, 5
+        >>> estimate = torch.stack([torch.ones(F, T), torch.zeros(F, T)])
+        >>> target = estimate[(1, 0), :, :]
+        >>> pit_loss_from_pair_wise(pair_wise_loss(estimate, target, axis=0), return_permutation=True)
+        (tensor(0.), array([1, 0]))
+
+        >>> K = 5
+        >>> estimate, target = torch.ones(K), torch.zeros(K)
+        >>> pit_loss_from_pair_wise(pair_wise_loss(estimate, target, axis=0))
+        tensor(1.)
+
+        >>> A, B, K, C, F = 4, 5, 3, 100, 128
+        >>> estimate, target = torch.ones(A, B, K, C, F), torch.zeros(A, B, K, C, F)
+        >>> pit_loss_from_pair_wise(pair_wise_loss(estimate, target, axis=-3))
+        tensor(1.)
+    """
+    # axis = axis % estimate.ndimension()
+    sources = estimate.size()[axis]
+    assert sources < 30, f'Are you sure? sources={sources}'
+    if loss_fn in [torch.nn.functional.cross_entropy]:
+        estimate_shape = list(estimate.shape)
+        del estimate_shape[1]
+        assert estimate_shape == list(target.shape), (
+            f'{estimate.shape} (N, K, ...) does not match {target.shape} (N, ...)'
+        )
+    else:
+        assert estimate.size() == target.size(), (
+            f'{estimate.size()} != {target.size()}'
+        )
+
+    pair_wise_loss = []
+
+    if estimate.ndim == target.ndim:
+        indexer_e = [slice(None), ] * estimate.ndim
+        indexer_t = [slice(None), ] * target.ndim
+        for i in range(sources):
+            indexer_e[axis] = i
+            for j in range(0, sources):
+                indexer_t[axis] = j
+                pair_wise_loss.append(loss_fn(
+                    estimate[tuple(indexer_e)],
+                    target[tuple(indexer_t)],
+                ))
+        pair_wise_loss = torch.stack(pair_wise_loss, 0).reshape(sources, sources)
+    else:
+        assert loss_fn == torch.nn.functional.cross_entropy, loss_fn
+        assert axis == 1, axis
+        import einops
+
+        pair_wise_loss = einops.reduce(torch.einsum(
+            'nc...,n...k->n...ck',
+            -torch.nn.LogSoftmax(dim=1)(estimate),
+            torch.nn.functional.one_hot(target, num_classes=sources).to(estimate.dtype)
+        ), 'n ... c k -> c k', reduction='mean')
+
+    return pair_wise_loss
+
+
+def pit_loss_from_pair_wise(
+        pair_wise_loss,
```

Comment:
Again, there already is a function called `pair_wise_loss`

Reply:
I changed it to `pair_wise_loss_matrix`.
I wanted to keep the name verbose (i.e. obvious connection to `pair_wise_loss` function).

Reply:
Maybe you can also change `pit_loss_from_pair_wise` to `pit_loss_from_loss_matrix`. I think this way the usage of this function is clearer.